proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+- *Union behaviors*: All union types support the following *union behaviors*:
+    - *Union conversions*: There are implicit *union conversions* from each case type to a union type.
+    - *Union matching*: Pattern matching against union values implicitly "unwraps" their contents, applying the pattern to the underlying value instead.
+    - *Union exhaustiveness*: Switch expressions over union values are exhaustive when all case types have been matched, without need for a fallback case.
+    - *Union nullability*: Nullability analysis has enhanced tracking of the null state of a union's contents.
+- *Union patterns*: All union types follow a basic *union pattern* (implement `IUnion` and offer public constructors for case types), but there are additional optional patterns that enable the union behaviors to be implemented in ways that may be more efficient.
+- *Union declarations*: A shorthand syntax allows declaration of union types directly. The implementation is "opinionated" - a struct declaration that follows the basic union pattern and stores the contents as a single reference field.
+- *Union interfaces*: A few interfaces are known by the language and used in its implementation of unions. `IUnion` is used to recognize union types and support union matching. An optional interface, `IUnion<TUnion>`, adds broader support for creating union values from potential case types, and is implemented by the union types arising from union declarations.
+
+## Motivation
+[motivation]: #motivation
+
+Unions are a long-requested C# feature, which allows expressing values from a closed set of types in a way that pattern matching can trust to be exhaustive.
+
+The separation between union *types* and union *declarations* allows C# to have a succinct union declaration syntax with opinionated semantics, while also allowing existing types or types with other implementation choices to opt into union behaviors.
+
+The proposed unions in C# are unions of *types* and not "discriminated" or "tagged". "Discriminated unions" can be expressed in terms of "type unions" by using fresh type declarations as case types.
+
+## Detailed design
+[design]: #detailed-design
+
+### Union interfaces
+
+The following interfaces are recognized by the language in its implementation of union features.
+
+#### Union access interface
+
+The `IUnion` interface marks a type as a union type at compile time and provides a way to access union contents at runtime.
+
+```csharp
+public interface IUnion
+{
+    // The value of the union or null
+    object? Value { get; }
+}
+```
+As we'll see later:
+* The `IUnion` interface is all that is necessary for a type to be considered a union by the language.  
+* The `Value` property of the interface is targeted by the compiler to implement pattern matching.
+* Union declarations are generated by the compiler to implement this interface.
```

Comment:
perhaps have a clause that mentions tehre may be alternate approaches that are more efficient, which would support non-boxing consumption, but that that will be discussed in seciton XXX.

Reply:
Tried this out, but I feel it draws away from this just being a non-normative summary of what the interface is for, so I removed it again. Happy to take suggestions though!